appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,178 @@
+    private static File junitJarFile;
+
+    @BeforeClass
+    public static void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        lock = new CyclicBarrier(3);
+        executor = Executors.newFixedThreadPool(60);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @AfterClass
+    public static void shutdown() {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+    }
+
+    @Test
+    public void check_findResourceInternalFromJars_thread_safety() throws Exception {
+        final ClassLoader classLoader = this.getClass().getClassLoader();
+        final WebappClassLoader webappClassLoader = new WebappClassLoader(classLoader, null);
+        webappClassLoader.start();
+        webappClassLoader.setResources(new FileDirContext());
+
+        CompletableFuture<Exception> result = new CompletableFuture<>();
+
+        // Create the tasks, and have them each run at the same time
+        // using the cyclic barrier
+        Runnable lookupTask = waitAndDo(lock, result, () -> lookup(classLoader, webappClassLoader));
+        Runnable addTask = waitAndDo(lock, result, () -> add(classLoader, webappClassLoader));
+        Runnable closeTask = waitAndDo(lock, result, () -> webappClassLoader.closeJARs(true));
+
+        try {
+            // Run the methods at the same time
+            for (int i = 0; i < 100; i++) {
+                executor.execute(addTask);
+                executor.execute(lookupTask);
+                executor.execute(closeTask);
+            }
+            // Check to see if any completed exceptionally
+            Exception ex = result.get(200, TimeUnit.MILLISECONDS);
+            if (ex != null) {
+                throw ex;
+            }
+        } catch (TimeoutException ex) {
+            // Success!
+        } finally {
+            webappClassLoader.close();
+        }
+    }
+
+    private void add(ClassLoader realClassLoader, WebappClassLoader webappClassLoader) throws IOException {
+        List<JarFile> jarFiles = findJarFiles(realClassLoader);
+
+        for (JarFile j : jarFiles) {
+            try {
+                webappClassLoader.addJar(junitJarFile.getName(), j, junitJarFile);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private void lookup(ClassLoader realClassLoader, WebappClassLoader webappClassLoader) throws Exception {
+        for (JarFile jarFile : findJarFiles(realClassLoader)) {
+            for (JarEntry entry : Collections.list(jarFile.entries())) {
+                webappClassLoader.findResource(entry.getName());
+                // System.out.println("Looked up " + resourceEntry);
+                Thread.sleep(0, 100);
```

Comment:
This causes slowing down of all lookup tasks. When I tried to run much more threads to cause deadlock, I replaced this with Thread.yield(). But it is not generally better, it depends on what we hunt.